crates/core/src/db/messages/write.rs

```diff
@@ -1,13 +1,52 @@
+use std::fmt;
+
+use anyhow::Context as _;
 pub use spacetimedb_lib::DataKey;
+use spacetimedb_sats::buffer::{BufReader, BufWriter, DecodeError};
+
+#[cfg(test)]
+use proptest::prelude::*;
+#[cfg(test)]
+use proptest_derive::Arbitrary;
 
-#[derive(Debug, Copy, Clone)]
+/// A single write operation within a [`super::transaction::Transaction`].
+///
+/// Encoding:
+///
+/// ```text
+/// <flags(1)><set_id(4)><value(1-33)>
+/// ```
+#[derive(Debug, Copy, Clone, PartialEq)]
+#[cfg_attr(test, derive(Arbitrary))]
 pub struct Write {
     pub operation: Operation,
-    pub set_id: u32,
+    pub set_id: u32, // aka table id
+    #[cfg_attr(test, proptest(strategy = "arbitrary::datakey()"))]
     pub data_key: DataKey,
 }
 
-#[derive(Debug, Copy, Clone)]
+#[cfg(test)]
+mod arbitrary {
+    use super::*;
+
+    pub fn datakey() -> impl Strategy<Value = DataKey> {
+        prop_oneof![
+            prop::collection::vec(any::<u8>(), 0..31).prop_map(DataKey::from_data),
```

Comment:
I'd comment on the reason for the two ranges.

Reply:
ea62cf8c779c1860207c645a8ff189c14c1b0f9b